infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuInterClusterReplicationServerNode.java

```diff
@@ -38,97 +44,169 @@ public class CorfuInterClusterReplicationServerNode implements AutoCloseable {
+     * @param localCorfuEndpoint
+     * @param localNodeId
+     * @param topologyConfigId Topology Configuration id
+     * @param sourceClientToMetadataManagerMap Map of remote(Source) cluster
+     * ids to the Metadata Manager of each
      */
-    public CorfuInterClusterReplicationServerNode(@Nonnull ServerContext serverContext,
-                                                  @Nonnull Map<Class, AbstractServer> serverMap) {
+    public CorfuInterClusterReplicationServerNode(
+        @Nonnull ServerContext serverContext,
+        @Nonnull LogReplicationConfig config, Set<String> remoteClusterIds,
+        String localCorfuEndpoint, String localNodeId, long topologyConfigId,
+        Map<String, LogReplicationMetadataManager> sourceClientToMetadataManagerMap) {
+
+        this.topologyConfigId = topologyConfigId;
         this.serverContext = serverContext;
-        this.serverMap = serverMap;
+        this.logReplicationConfig = config;
+
+        logReplicationServer = new LogReplicationServer(serverContext,
+            localNodeId, logReplicationConfig, remoteClusterIds,
+            localCorfuEndpoint, topologyConfigId, sourceClientToMetadataManagerMap);
+
+        this.serverMap = ImmutableMap.<Class, AbstractServer>builder()
+            .put(BaseServer.class, new BaseServer(serverContext))
+            .put(LogReplicationServer.class, logReplicationServer)
+            .build();
 
         this.close = new AtomicBoolean(false);
-        this.router = new LogReplicationServerRouter(new ArrayList<>(serverMap.values()));
+        this.router = new LogReplicationServerRouter(
+            new ArrayList<>(serverMap.values()));
         this.serverContext.setServerRouter(router);
+
+        logReplicationServerRunner = Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder().setNameFormat("replication-server-runner")
+                .build());
+
+        // Start and listen to the server immediately without delay
+        logReplicationServerFuture =
+            logReplicationServerRunner.schedule(new LogReplicationServerRunner(),
+                0, TimeUnit.NANOSECONDS);
     }
 
     /**
      * Wait on Corfu Server Channel until it closes.
      */
-    public void startAndListen() {
+    private void startAndListen() {
         try {
             log.info("Starting server transport adapter...");
             router.getServerAdapter().start().get();
-        } catch (Exception e) {
-            throw new UnrecoverableCorfuError(e);
+        } catch (InterruptedException e) {
+            // The server can be interrupted and stopped on a role switch.
+            // It should not be treated as fatal
+            log.warn("Server interrupted.  It could be due to a role switch");
+        } catch (Throwable th) {
+            log.error("LogReplicationServer exiting due to unrecoverable error:",
+                th);
+            System.exit(EXIT_ERROR_CODE);
```

Comment:
Yes, because we register a shutdown hook on JVM startup -
https://github.com/CorfuDB/CorfuDB/blob/master/infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuInterClusterReplicationServer.java#L251

This will cleanly shutdown the discovery service and also call flushAsyncLogAppender().  Verified by testing.